pandas/plotting/_matplotlib/hist.py

```diff
@@ -19,19 +24,35 @@ def __init__(self, data, bins=10, bottom=0, **kwargs):
         MPLPlot.__init__(self, data, **kwargs)
 
     def _args_adjust(self):
+
+        # calculate bin number separately in different subplots
+        # where subplots are created based on by argument
         if is_integer(self.bins):
-            # create common bin edge
-            values = self.data._convert(datetime=True)._get_numeric_data()
-            values = np.ravel(values)
-            values = values[~isna(values)]
+            if self.by is None:
+                self.bins = self._calculate_bins(self.data)
 
-            _, self.bins = np.histogram(
-                values, bins=self.bins, range=self.kwds.get("range", None)
-            )
+            else:
+                grouped = self.data.groupby(self.by)[self.columns]
+                bins_list = []
+                for key, group in grouped:
+                    bins_list.append(self._calculate_bins(group))
+                self.bins = bins_list
 
         if is_list_like(self.bottom):
             self.bottom = np.array(self.bottom)
 
+    def _calculate_bins(self, data: DataFrame) -> np.array:
```

Comment:
maybe a module level function?

Reply:
emm, i'd prefer to leave this here, since it's quite related only to `hist` plots, and it's used as part of post-processing for `self.bins` for hist plots. Moreover, inside the function, it calls some other `self` variables, so i would actually prefer to leave as is.

what do you think?